Display values in info box only if value is interesting

sfuchs79 opened the pull request:
<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [x] Bug fix
- [x] Functional change
- [ ] New feature
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
<!-- Describe your pull request in detail. -->
Valid flag introduced for some of the defined types.
Valid flag used for compass bearing and NDL.
Here big ? on my side if this is acceptable/a good idea or not.

Display bearing only if a bearing was set in the DC based on valid flag.
Some DCs seem to return -1 if no bearing is set.

Display NDL based on the valid flag.

Display pXX, EAD, END, density, MOD only if values are larger than 0.

In profile don't display data from two first and two last plot_data
entries in info box.

Plus one typo corrected which broke a translation of string "density:".

### Changes made:
<!-- Enumerate the changes with 1), 2), 3) etc. -->
<!-- Ensure the test cases are updated if needed. -->

### Related issues:
<!-- Reference issues with #<issue-num>. -->
<!-- Write "Fixes #<issue-num" to notify Github that this PR fixes an issue. -->
This is my proposal for #692 

### Additional information:
<!-- Include sample dive log or other relevant information to allow testing the change where feasible. -->

### Mentions:
<!-- Mention users that you want to review your pull request with @<user-name>. Leave empty if not sure. -->

dirkhh commented on core/units.h:
```diff
@@ -96,6 +100,7 @@ typedef struct
+	int valid;
 } bearing_t; // compass bearing
 
 typedef struct
```
Not a fan of these 'valid' members. Simply set these to a known invalid value (and create a macro DURATION_VALID() and BEARING_VALID() that checks for that value).

sfuchs79 replied:
Ok for me and I also won't do this one in a hurry.
And I please need some guidance for a better solution.

In general I would prefer to keep a value '0' being the "flag" for not initialized as default because this works most of the time.
For some items like bearing or ndl we need s.th. else. I would prefer '-1'. Would it be ok to change duration_t from uint to int?
And using '-1' (or anything else != 0) also means that we need to take care to init this in some places. Currently we use e.g. a calloc to init the memory for a new sample to all '0'. We would need to set some specific '-1' then. But don't worry, that's not more than a few lines of code. On the other hand '-1' will make the data transfer from samples to plot_data easier again.

sfuchs79 commented:
Pushed an alternative solution for review. Now w/o "valid" flags but using '-1' as "invalid" marker for NDL and bearing.

sfuchs79 commented:
@dirkhh 
Short update: This one now would need some feedback if the variant with using '-1' as flag for uninitialized is ok in general. I will be diving during the next days and won't be able to change and test but I would be available for discussion.

I for sure recognized that TestParse fails for a specific DC file. I already investigated this and can explain the why. But discussing this is a second step after deciding if the change is acceptable in general.

dirkhh commented:
I think this is much better than the first attempt. But I'd love to see @torvalds opinion on this as well.

torvalds commented:
I'm certainly ok with this.

The "bearing" thing is not well-defined, though: the comment says -32k to +32k degrees, but then -1 means "no bearing". So maybe the comment should be updated to 0-360 degrees? 

But that's a nit, and not really all that new. The old comment made little sense anyway.

dirkhh commented:
OK, next steps:
a) address @torvalds' comment regarding the comment in the code :-)
b) address the test failure (and make sure there's no regression with old files)

sfuchs79 commented:
Ok, thanks for the positive feedback.

Regarding the bearing: Are we sure that we will always get 0-360° from the DCs? Then I will update the comment.

About the failed test:
Here I tend to blame the Seabear file format or the way we import it. You please need to have a look at an example file for explanation:
https://github.com/Subsurface-divelog/subsurface/blob/master/dives/TestDiveSeabearH3_0002SBDT.CSV#L23

The Seabear file (and maybe also other DCs?!) is also bad in distinguishing between no NDL info available and NDL=0. There is a first entry with NDL=0 which means "no NDL", then there are entries with NDL>>0 and then there may be real NDL=0 entries.
As long as Subsurface was also not distinguishing between the two cases we were "on the same level" and everything was "ok". If we improve Subsurface we need to decide how to handle the file(s). Two options:
- Ignore/accept the "wrong" entry and update our reference XML files.
- Make our import more intelligent and translate the first NDL=0 into our new internal '-1'. This could be done easily by also looking at the TTS. But I was not able to figure out yet where to do this in our code.

BTW: No real code and test updates from my side before mid of next week - this time I'm diving :-)

sfuchs79 commented:
Found more items for the ToDo list for this item:
- Issue with NDL=0 at the very beginning of the dive is also present when importing from my OSTC. So the item above that we need to translate a NDL=0 at the very beginning of a dive into a '-1' maybe needs to be applied everywhere.
- We today seem to add one last sample for many dives but this sample today already has incorrect data. It's not only a new thing that the NDL is set to 0 instead of value from last sample but also e.g. the CNS is set to 0 and then e.g. exported to the XML as "cns='0%'" while the dive had a CNS info>>0 in some samples before.
@dirkhh : Where does this last sample come from?
- In the export to XML and ... better change all %u to %d
- For dives w/o samples I also need to correct s.th.

sfuchs79 commented:
@dirkhh @torvalds 

Ok, a little progress: I updated this and fixed the comments (Linus hint) and the issue with the manually entered dives. And I structured the commits in a better way.

What is now still open is explained already above and is the thing what also causes the TestParse to fail.
But I better explain again with another example which is importing 3 dives from my OSTC. In fact I think that the issue is present for every import (and export?):

- If I readout 3 example dives from my OSTC to a new and empty logfile and then save to an Subsurface XML I see two incorrect sample entries for every dive: The very first one and the very last one.
[New_Readout.zip](https://github.com/Subsurface-divelog/subsurface/files/1444176/New_Readout.zip)

- The first one always has NDL=0 info and sometimes also some incorrect CNS info. The last one always has NDL=-1 info plus incorrect CNS info plus sometimes some other incorrect info.

- I can explain the incorrect NDL=0 info in very first sample but I'm currently not able to fix it.
Where could we correct this in a global way saying: NDL=0 in first sample can not exist and will be translated into '-1'?

- I even can't explain the incorrect CNS entries and the NDL in last sample. It seems to me that there is a last sample created for which the info is not correctly copied from the sample before. Is this even a very old bug? Maybe even in libdc?

I'll also drop Linus a mail and ask for help :-)

torvalds commented:
Ok, so that

    ndl='71582788:15 min'

is because 71582788:15 is 4294967295 seconds.

So something treated (int)-1 as a unsigned number, and decided to print it out as minutes/seconds instead of treating it as "no ndl".

So you need to check core/save-xml.c (and core/save-git.c) and the loading path, and decide what you do with "no NDL" vs "zero NDL". 

Right now the saving code does

        /* the deco/ndl values are stored whenever they change */
        if (sample->ndl.seconds != old->ndl.seconds) {
                put_format(b, " ndl=%u:%02u", FRACTION(sample->ndl.seconds, 60));
                old->ndl = sample->ndl;
        }

and that is wrong for the "-1" case. So you need to think about what is going on, and why only that _last_ sample had a NDL of -1, while apparently the previous one had a zero one or something.

sfuchs79 commented:
Yes, for me currently the key question is why the very last sample even has a NDL of '-1' and not the very same NDL info of the sample one before. And same question for CNS,...
And this is what I'm currently not able to figure out. Where is this last sample created? I somehow have the feeling this is a kind of "special" last sample and we up to now did a mistake when creating it.

sfuchs79 commented:
This ndl/compass bearing/info box thing is now in an acceptable state for final review and merge.
Test pass now. I even didn't change that much any more.
What really changed is my understanding of some small abnormalities in cns/ndl/stop time entries in some test reference and my own (OSTC) xml files. In discussions with @torvalds and @janmulder it became clear that these are separate issues and have nothing to do with this change being right or wrong.
See https://github.com/Subsurface-divelog/libdc/pull/10

janmulder commented:
On the current master (a8d4ff308d41987) this PR introduces ndl='71582788:15 min in the mysterious last sample (OSTC).

sfuchs79 commented:
Hmmm, not good. I thought this didn't happen any more in my tests because the internal sample NDL won't go back to '-1' after it once is at a value >=0. And also the tests didn't fail any longer.
I have to tests this again.
But as you are saying the last sample in data from OSTC is still a strange thing at it's own.

sfuchs79 commented:
Yes, confirmed. Happens also here.
I still think the issue is that somewhere a last sample is created which doesn't inherit it's data correctly from the previous sample(s) but has all data reset to initial values.
That leads to incorrect:
- NDL="-1" - yeah that this is printed as "ndl='71582788:15 min'" is my fault - thought we would never run into this case...
- cns='0%' for any dive which had cns>0 in a sample before
- stoptime='0:00 min' for any dive which had stoptime>0 in a sample before

But where is this incorrect sample created?

Line 586:   <sample time='93:10 min' depth='0.01 m' ndl='71582788:15 min' stoptime='0:00 min' cns='0%' />
	Line 914:   <sample time='51:20 min' depth='0.02 m' ndl='71582788:15 min' />
	Line 1513:   <sample time='96:20 min' depth='0.0 m' ndl='71582788:15 min' cns='0%' />
	Line 1918:   <sample time='64:10 min' depth='0.0 m' ndl='71582788:15 min' />
	Line 2650:   <sample time='118:40 min' depth='0.01 m' ndl='71582788:15 min' cns='0%' />
	Line 3255:   <sample time='97:30 min' depth='0.17 m' ndl='71582788:15 min' cns='0%' />
	Line 3526:   <sample time='41:50 min' depth='0.06 m' ndl='71582788:15 min' cns='0%' />
	Line 3794:   <sample time='41:10 min' depth='0.37 m' ndl='71582788:15 min' cns='0%' />
	Line 5118:   <sample time='94:20 min' depth='0.01 m' ndl='71582788:15 min' stoptime='0:00 min' cns='0%' />
	Line 5512:   <sample time='62:20 min' depth='0.14 m' ndl='71582788:15 min' cns='0%' />
	Line 5909:   <sample time='62:50 min' depth='0.01 m' ndl='71582788:15 min' cns='0%' />
	Line 6293:   <sample time='60:40 min' depth='0.0 m' ndl='71582788:15 min' stoptime='0:00 min' cns='0%' />
	Line 6840:   <sample time='87:50 min' depth='0.0 m' ndl='71582788:15 min' stoptime='0:00 min' cns='0%' />
	Line 7448:   <sample time='96:20 min' depth='0.0 m' ndl='71582788:15 min' stoptime='0:00 min' cns='0%' />
	Line 7849:   <sample time='11:10 min' depth='0.37 m' ndl='71582788:15 min' />
	Line 8236:   <sample time='60:50 min' depth='0.01 m' ndl='71582788:15 min' cns='0%' />
	Line 8957:   <sample time='116:50 min' depth='0.0 m' ndl='71582788:15 min' cns='0%' />
	Line 9993:   <sample time='95:50 min' depth='0.0 m' ndl='71582788:15 min' cns='0%' />
	Line 10466:   <sample time='75:30 min' depth='0.01 m' ndl='71582788:15 min' stoptime='0:00 min' cns='0%' />
	Line 11058:   <sample time='94:50 min' depth='0.0 m' ndl='71582788:15 min' stoptime='0:00 min' cns='0%' />
	Line 11488:   <sample time='68:20 min' depth='0.0 m' ndl='71582788:15 min' stoptime='0:00 min' cns='0%' />
	Line 11906:   <sample time='66:20 min' depth='0.0 m' ndl='71582788:15 min' stoptime='0:00 min' cns='0%' />
	Line 12766:   <sample time='139:30 min' depth='0.01 m' ndl='71582788:15 min' cns='0%' />
	Line 13600:   <sample time='135:10 min' depth='0.11 m' ndl='71582788:15 min' cns='0%' />
	Line 14026:   <sample time='67:20 min' depth='0.02 m' ndl='71582788:15 min' stoptime='0:00 min' cns='0%' />
	Line 14596:   <sample time='90:50 min' depth='0.02 m' ndl='71582788:15 min' stoptime='0:00 min' cns='0%' />
	Line 15198:   <sample time='96:30 min' depth='0.15 m' ndl='71582788:15 min' stoptime='0:00 min' cns='0%' />
	Line 15905:   <sample time='113:20 min' depth='0.0 m' ndl='71582788:15 min' stoptime='0:00 min' cns='0%' />
	Line 16325:   <sample time='66:40 min' depth='0.0 m' ndl='71582788:15 min' cns='0%' />
	Line 16808:   <sample time='77:10 min' depth='0.0 m' ndl='71582788:15 min' />

sfuchs79 commented:
@dirkhh @torvalds @janmulder 
I think the reason for the strange data in every last sample is here:
https://github.com/Subsurface-divelog/subsurface/blob/master/core/libdivecomputer.c#L332

We care here about the sticky values always for the previous sample as soon as we create a new sample.
But the last sample never has a following sample so this code will never be executed for the last sample. So the last sample always may have initial values for the list of values there in the code (in_deco, ndl, stoptime,...) and this is not correct.

Can we easily change this by copying the values for the current sample immediately?

sfuchs79 commented:
I wonder if we can fix it that easy:
#812

janmulder commented:
@sfuchs79 that's a very sharp and correct observation. Further comment on #812
